Modules/_csv.c

```diff
@@ -1110,197 +1103,126 @@ csv_reader(PyObject *module, PyObject *args, PyObject *keyword_args)
-    if (!join_check_rec_size(self, rec_len))
-        return 0;
-
-    self->rec_len = join_append_data(self, field_kind, field_data, field_len,
-                                     &quoted, 1);
-    self->num_fields++;
 
-    return 1;
-}
-
-static int
-join_append_lineterminator(WriterObj *self)
-{
-    Py_ssize_t terminator_len, i;
-    int term_kind;
-    const void *term_data;
+    /* open if needed */
+    if (*quoted) {
+        if (PyUnicodeWriter_WriteChar(writer, dialect->quotechar) < 0) {
+            return -1;
+        }
+    }
 
-    terminator_len = PyUnicode_GET_LENGTH(self->dialect->lineterminator);
-    if (terminator_len == -1)
-        return 0;
+    if (!is_none && field_len > 0) {
+        Py_ssize_t i;
+        for (i = 0; i < field_len; i++) {
+            Py_UCS4 c = PyUnicode_READ_CHAR(field, i);
+
+            if (dialect->quoting == QUOTE_NONE) {
+                /* escape structural characters when we cannot quote. */
+                if (_is_structural_char(c, dialect, term_len) ||
+                    c == dialect->escapechar ||
+                    c == dialect->quotechar) {
+                    if (dialect->escapechar == NOT_SET) {
+                        PyErr_SetString(self->error_obj, "need to escape, but no escapechar set");
+                        return -1;
+                    }
+                    if (PyUnicodeWriter_WriteChar(writer, dialect->escapechar) < 0) {
+                        return -1;
+                    }
+                }
+            }
+            else {
+                /* handle in-field quoting/escaping. */
+                if (c == dialect->quotechar) {
+                    if (dialect->doublequote) {
+                        /* double the quote inside a quoted field */
+                        if (PyUnicodeWriter_WriteChar(writer, dialect->quotechar) < 0) {
+                            return -1;
+                        }
+                    }
+                    else {
+                        /* ...or escape if we cannot double. */
+                        if (dialect->escapechar == NOT_SET) {
+                            PyErr_SetString(self->error_obj, "need to escape, but no escapechar set");
+                            return -1;
+                        }
+                        if (PyUnicodeWriter_WriteChar(writer, dialect->escapechar) < 0) {
+                            return -1;
+                        }
+                    }
+                }
+                else if (c == dialect->escapechar && dialect->escapechar != NOT_SET) {
+                    /* escape literal escapechar */
+                    if (PyUnicodeWriter_WriteChar(writer, dialect->escapechar) < 0) {
+                        return -1;
+                    }
+                }
+            }
 
-    /* grow record buffer if necessary */
-    if (!join_check_rec_size(self, self->rec_len + terminator_len))
-        return 0;
+            if (PyUnicodeWriter_WriteChar(writer, c) < 0) {
```

Comment:
It would be interesting to try calling WriteSubstring() at once rather than writing characters one by one.